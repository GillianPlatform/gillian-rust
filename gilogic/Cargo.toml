[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }