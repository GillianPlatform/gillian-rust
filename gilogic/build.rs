fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
