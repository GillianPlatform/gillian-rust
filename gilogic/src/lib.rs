#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
