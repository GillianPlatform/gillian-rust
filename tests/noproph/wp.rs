//@check-pass
extern crate gilogic;

use gilogic::{
    macros::{with_freeze_lemma, *},
    Ownable,
};

struct WP<T> {
    x: *mut N<T>,
    y: *mut N<T>,
}

struct N<T> {
    v: T,
    o: *mut N<T>,
}

#[predicate]
fn wp<T: Ownable>(wp: In<WP<T>>, x: *mut N<T>, y: *mut N<T>) {
    assertion!(|v_x, v_y|
        (wp == WP { x, y }) *
        (x -> N { v: v_x, o: y }) *
        (y -> N { v: v_y, o: x }) *
        v_x.own() * v_y.own()
    )
}

#[extract_lemma(
    forall x, y.
    from { wp_ref_mut_xy(p, x, y) }
    extract { <&mut T as Ownable>::own(&mut (*x).v) }
)]
fn extract_x<'a, T: Ownable>(p: &'a mut WP<T>);

#[extract_lemma(
    forall x, y.
    from { wp_ref_mut_xy(p, x, y) }
    extract { <&mut T as Ownable>::own(&mut (*y).v) }
)]
fn extract_y<'a, T: Ownable>(p: &'a mut WP<T>);

#[with_freeze_lemma(
    lemma_name = freeze_xy,
    predicate_name = wp_ref_mut_xy,
    frozen_variables = [x, y],
)]
impl<T: Ownable> Ownable for WP<T> {
    #[predicate]
    fn own(self) {
        assertion!(|x: *mut N<T>, y: *mut N<T>| wp(self, x, y))
    }
}

impl<T: Ownable> WP<T> {
    #[show_safety]
    fn new(x: T, y: T) -> Self {
        let xbox = Box::new(N {
            v: x,
            o: std::ptr::null_mut(),
        });
        let ybox = Box::new(N {
            v: y,
            o: std::ptr::null_mut(),
        });

        let xptr = Box::leak(xbox) as *mut N<T>;
        let yptr = Box::leak(ybox) as *mut N<T>;

        unsafe {
            (*yptr).o = xptr;
            (*xptr).o = yptr;
        }
        WP { x: xptr, y: yptr }
    }

    #[show_safety]
    fn assign_first<'a>(&'a mut self, x: T) {
        unsafe {
            (*self.x).v = x;
        }
    }

    #[show_safety]
    fn first_mut<'a>(&'a mut self) -> &'a mut T {
        unsafe {
            freeze_xy(self);
            let ret = &mut (*self.x).v;
            extract_x(self);
            ret
        }
    }

    #[show_safety]
    fn second_mut<'a>(&'a mut self) -> &'a mut T {
        unsafe {
            freeze_xy(self);
            let ret = &mut (*self.y).v;
            extract_y(self);
            ret
        }
    }
}
