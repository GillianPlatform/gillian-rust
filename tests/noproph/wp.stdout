#begin_init_data
{"Global":["Struct","ReprRust",[]],"N":["Struct","ReprRust",[["v",["Param",0]],["o",["Ptr",{"mut":true,"ty":["Adt",["N",[["Param",0]]]]}]]]],"WP":["Struct","ReprRust",[["x",["Ptr",{"mut":true,"ty":["Adt",["N",[["Param",0]]]]}]],["y",["Ptr",{"mut":true,"ty":["Adt",["N",[["Param",0]]]]}]]]]}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_no_pcy.gil", "i__lang_no_pcy.gil";

axiomatic spec extract_x(pLft_a, pty_T0, p)
[[ "wp_ref_mut_xy::<T>"(#pLft_a, #pty_T0, #p, #x, #y) * <lft>(#pLft_a; true)
  * types(#lvar_4: Obj, #lvar_5: List) * (#y == {{ #lvar_4, #lvar_5 }})
  * types(#lvar_2: Obj, #lvar_3: List) * (#x == {{ #lvar_2, #lvar_3 }})
  * types(#lvar_0: Obj, #lvar_1: List) * (#p == {{ #lvar_0, #lvar_1 }})
  * (p == #p)
  * (pty_T0 == #pty_T0)
  * (pLft_a == #pLft_a) ]]
[[ "<&mut T as gilogic::Ownable>::own"(#pLft_a, #pty_T0, {{ l-nth(#x, 0i), l+(l-nth(#x, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ #pty_T0 }} }} }} }}) }})
  * <lft>(#pLft_a; true) ]]
normal

axiomatic spec extract_y(pLft_a, pty_T0, p)
[[ "wp_ref_mut_xy::<T>"(#pLft_a, #pty_T0, #p, #x, #y) * <lft>(#pLft_a; true)
  * types(#lvar_4: Obj, #lvar_5: List) * (#y == {{ #lvar_4, #lvar_5 }})
  * types(#lvar_2: Obj, #lvar_3: List) * (#x == {{ #lvar_2, #lvar_3 }})
  * types(#lvar_0: Obj, #lvar_1: List) * (#p == {{ #lvar_0, #lvar_1 }})
  * (p == #p)
  * (pty_T0 == #pty_T0)
  * (pLft_a == #pLft_a) ]]
[[ "<&mut T as gilogic::Ownable>::own"(#pLft_a, #pty_T0, {{ l-nth(#y, 0i), l+(l-nth(#y, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ #pty_T0 }} }} }} }}) }})
  * <lft>(#pLft_a; true) ]]
normal

axiomatic spec "extract_x::<T>"(pLft_a, pty_T0, p)
[[ "wp_ref_mut_xy::<T>"(#pLft_a, #pty_T0, #p, #x, #y) * <lft>(#pLft_a; true)
  * types(#lvar_4: Obj, #lvar_5: List) * (#y == {{ #lvar_4, #lvar_5 }})
  * types(#lvar_2: Obj, #lvar_3: List) * (#x == {{ #lvar_2, #lvar_3 }})
  * types(#lvar_0: Obj, #lvar_1: List) * (#p == {{ #lvar_0, #lvar_1 }})
  * (p == #p)
  * (pty_T0 == #pty_T0)
  * (pLft_a == #pLft_a) ]]
[[ "<&mut T as gilogic::Ownable>::own"(#pLft_a, #pty_T0, {{ l-nth(#x, 0i), l+(l-nth(#x, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ #pty_T0 }} }} }} }}) }})
  * <lft>(#pLft_a; true) ]]
normal

axiomatic spec "extract_y::<T>"(pLft_a, pty_T0, p)
[[ "wp_ref_mut_xy::<T>"(#pLft_a, #pty_T0, #p, #x, #y) * <lft>(#pLft_a; true)
  * types(#lvar_4: Obj, #lvar_5: List) * (#y == {{ #lvar_4, #lvar_5 }})
  * types(#lvar_2: Obj, #lvar_3: List) * (#x == {{ #lvar_2, #lvar_3 }})
  * types(#lvar_0: Obj, #lvar_1: List) * (#p == {{ #lvar_0, #lvar_1 }})
  * (p == #p)
  * (pty_T0 == #pty_T0)
  * (pLft_a == #pLft_a) ]]
[[ "<&mut T as gilogic::Ownable>::own"(#pLft_a, #pty_T0, {{ l-nth(#y, 0i), l+(l-nth(#y, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ #pty_T0 }} }} }} }}) }})
  * <lft>(#pLft_a; true) ]]
normal

pred "wp::<T>"(+pty_T0, +wp, x, y):
  (#wp == {{ #x, #y }})
  * <value>(l-nth(#x, 0i), l-nth(#x, 1i), {{ "adt", "N", {{ pty_T0 }} }}; {{ #v_x, #y }})
  * <value>(l-nth(#y, 0i), l-nth(#y, 1i), {{ "adt", "N", {{ pty_T0 }} }}; {{ #v_y, #x }})
  * "<T as gilogic::Ownable>::own"(pty_T0, #v_x)
  * "<T as gilogic::Ownable>::own"(pty_T0, #v_y)
  * types(#lvar_0: Obj, #lvar_1: List) * (#x == {{ #lvar_0, #lvar_1 }})
  * types(#lvar_2: Obj, #lvar_3: List) * (#y == {{ #lvar_2, #lvar_3 }})
  * (y == #y)
  * (x == #x)
  * (wp == #wp);


pred "wp_ref_mut_xy::<T>"(+pLft_0, +pty_T0, +REFERENCE, x, y):
  "wp::<T>"(pty_T0, #SELF, #x, #y)
  * <value>(l-nth(#REFERENCE, 0i), l-nth(#REFERENCE, 1i), {{ "adt", "WP", {{ pty_T0 }} }}; #SELF)
  * types(#lvar_6: Obj, #lvar_7: List) * (#REFERENCE == {{ #lvar_6, #lvar_7 }})
  * types(#lvar_8: Obj, #lvar_9: List) * (#x == {{ #lvar_8, #lvar_9 }})
  * types(#lvar_10: Obj, #lvar_11: List) * (#y == {{ #lvar_10, #lvar_11 }})
  * (y == #y)
  * (x == #x)
  * (REFERENCE == #REFERENCE);
guard: <lft>(pLft_0; true);

pred "<WP<T> as gilogic::Ownable>::own"(+pty_T0, +self):
  "wp::<T>"(pty_T0, #self, #x, #y) * (self == #self);


abstract pred "<T as gilogic::Ownable>::own"(+pty_T0, +self);


pred "<&mut T as gilogic::Ownable>::own"(+pLft_erased, +pty_T0, +self):
  <value>(l-nth(#self, 0i), l-nth(#self, 1i), pty_T0; #v)
  * "<T as gilogic::Ownable>::own"(pty_T0, #v)
  * types(#lvar_0: Obj, #lvar_1: List) * (#self == {{ #lvar_0, #lvar_1 }})
  * (self == #self);
guard: <lft>(pLft_erased; true);

pred "<&mut WP<T> as gilogic::Ownable>::own"(+pLft_erased, +pty_T0, +self):
  <value>(l-nth(#self, 0i), l-nth(#self, 1i), {{ "adt", "WP", {{ pty_T0 }} }}; #v)
  * "<WP<T> as gilogic::Ownable>::own"(pty_T0, #v)
  * types(#lvar_0: Obj, #lvar_1: List) * (#self == {{ #lvar_0, #lvar_1 }})
  * (self == #self);
guard: <lft>(pLft_erased; true);

pred "<() as gilogic::Ownable>::own"(+self):
  (#self == {{  }}) * (self == #self);


lemma "freeze_xy::<T>"(pLft_0, pty_T0, REFERENCE)[[
  "<&mut WP<T> as gilogic::Ownable>::own"(#pLft_0, #pty_T0, #REFERENCE)
  * <lft>(#pLft_0; true)
  * types(#lvar_4: Obj, #lvar_5: List) * (#REFERENCE == {{ #lvar_4, #lvar_5 }})
  * (REFERENCE == #REFERENCE)
  * (pty_T0 == #pty_T0)
  * (pLft_0 == #pLft_0)
]]
[[
  "wp_ref_mut_xy::<T>"(#pLft_0, #pty_T0, #REFERENCE, #x, #y)
  * <lft>(#pLft_0; true)
]]

spec "WP::<T>::new"(pty_T0, x, y)
[[ "<T as gilogic::Ownable>::own"(#pty_T0, #x)
  * "<T as gilogic::Ownable>::own"(#pty_T0, #y)
  * (y == #y)
  * (x == #x)
  * (pty_T0 == #pty_T0) ]]
[[ "<WP<T> as gilogic::Ownable>::own"(#pty_T0, ret) ]]
normal

proc "WP::<T>::new"(pty_T0, x, y) {
        m_1x := x;
        m_2y := y;
        m_11xptr := [alloc]({{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        m_12 := [alloc]({{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        m_14yptr := [alloc]({{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        m_15 := [alloc]({{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }});
  bb0:  g_0 := m_1x;
        m_5 := g_0;
        g_1 := "std::ptr::null_mut"({{ "adt", "N", {{ pty_T0 }} }});
        m_6 := g_1;
        goto bb1;
  bb1:  g_2 := m_5;
        g_3 := m_6;
        m_4 := {{ g_2, g_3 }};
        goto bb2;
  bb2:  g_4 := m_4;
        g_5 := "std::boxed::Box::<T>::new"({{ "adt", "N", {{ pty_T0 }} }}, g_4);
        m_3xbox := g_5;
        goto bb3;
  bb3:  g_6 := m_2y;
        m_9 := g_6;
        g_7 := "std::ptr::null_mut"({{ "adt", "N", {{ pty_T0 }} }});
        m_10 := g_7;
        goto bb4;
  bb4:  g_8 := m_9;
        g_9 := m_10;
        m_8 := {{ g_8, g_9 }};
        goto bb5;
  bb5:  g_10 := m_8;
        g_11 := "std::boxed::Box::<T>::new"({{ "adt", "N", {{ pty_T0 }} }}, g_10);
        m_7ybox := g_11;
        goto bb6;
  bb6:  g_12 := m_3xbox;
        m_13 := g_12;
        g_13 := [new_lft]();
        pLft_4 := l-nth(g_13, 0i);
        g_14 := m_13;
        g_15 := "std::boxed::Box::<T, A>::leak"(pLft_4, {{ "adt", "N", {{ pty_T0 }} }}, {{ "adt", "Global", {{  }} }}, g_14);
        u := [store_value](l-nth(m_12, 0i), l-nth(m_12, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_15);
        goto bb7;
  bb7:  g_17 := [load_value](l-nth(m_12, 0i), l-nth(m_12, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_16 := l-nth(g_17, 0i);
        u := [store_value](l-nth(m_11xptr, 0i), l-nth(m_11xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_16);
        g_18 := m_7ybox;
        m_16 := g_18;
        g_19 := [new_lft]();
        pLft_5 := l-nth(g_19, 0i);
        g_20 := m_16;
        g_21 := "std::boxed::Box::<T, A>::leak"(pLft_5, {{ "adt", "N", {{ pty_T0 }} }}, {{ "adt", "Global", {{  }} }}, g_20);
        u := [store_value](l-nth(m_15, 0i), l-nth(m_15, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_21);
        goto bb8;
  bb8:  g_23 := [load_value](l-nth(m_15, 0i), l-nth(m_15, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_22 := l-nth(g_23, 0i);
        u := [store_value](l-nth(m_14yptr, 0i), l-nth(m_14yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_22);
        g_25 := [load_value](l-nth(m_11xptr, 0i), l-nth(m_11xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_24 := l-nth(g_25, 0i);
        m_18 := g_24;
        g_26 := m_18;
        g_28 := [load_value](l-nth(m_14yptr, 0i), l-nth(m_14yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_27 := l-nth(g_28, 0i);
        u := [store_value](l-nth(g_27, 0i), l+(l-nth(g_27, 1i), {{ {{ "f", 1i, {{ "adt", "N", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_26);
        g_30 := [load_value](l-nth(m_14yptr, 0i), l-nth(m_14yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_29 := l-nth(g_30, 0i);
        m_19 := g_29;
        g_31 := m_19;
        g_33 := [load_value](l-nth(m_11xptr, 0i), l-nth(m_11xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_32 := l-nth(g_33, 0i);
        u := [store_value](l-nth(g_32, 0i), l+(l-nth(g_32, 1i), {{ {{ "f", 1i, {{ "adt", "N", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, g_31);
        m_17 := {{}};
        g_35 := [load_value](l-nth(m_11xptr, 0i), l-nth(m_11xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_34 := l-nth(g_35, 0i);
        m_20 := g_34;
        g_37 := [load_value](l-nth(m_14yptr, 0i), l-nth(m_14yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_36 := l-nth(g_37, 0i);
        m_21 := g_36;
        g_38 := m_20;
        g_39 := m_21;
        m_0 := {{ g_38, g_39 }};
        goto bb9;
  bb9:  goto bb10;
  bb10: goto bb11;
  bb11: goto bb12;
  bb12: g_40 := m_0;
        ret := g_40;
        goto rlab;
  rlab: u := [free](l-nth(m_11xptr, 0i), l-nth(m_11xptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_12, 0i), l-nth(m_12, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_14yptr, 0i), l-nth(m_14yptr, 1i), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_15, 0i), l-nth(m_15, 1i), {{ "ref", true, {{ "adt", "N", {{ pty_T0 }} }} }});
        return
};

spec "WP::<T>::assign_first"(pLft_a, pty_T0, self, x)
[[ "<&mut WP<T> as gilogic::Ownable>::own"(#pLft_a, #pty_T0, #self)
  * "<T as gilogic::Ownable>::own"(#pty_T0, #x)
  * <lft>(#pLft_a; true)
  * types(#lvar_0: Obj, #lvar_1: List) * (#self == {{ #lvar_0, #lvar_1 }})
  * (x == #x)
  * (self == #self)
  * (pty_T0 == #pty_T0)
  * (pLft_a == #pLft_a) ]]
[[ "<() as gilogic::Ownable>::own"(ret) * <lft>(#pLft_a; true) ]]
normal

proc "WP::<T>::assign_first"(pLft_a, pty_T0, self, x) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, g_0);
        m_2x := x;
  bb0:  g_1 := m_2x;
        m_3 := g_1;
        goto bb1;
  bb1:  g_2 := m_3;
        g_4 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_3 := l-nth(g_4, 0i);
        g_6 := [load_value](l-nth(g_3, 0i), l+(l-nth(g_3, 1i), {{ {{ "f", 0i, {{ "adt", "WP", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_5 := l-nth(g_6, 0i);
        u := [store_value](l-nth(g_5, 0i), l+(l-nth(g_5, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ pty_T0 }} }} }} }}), pty_T0, g_2);
        goto bb3;
  bb3:  m_0 := {{}};
        goto bb4;
  bb4:  g_7 := m_0;
        ret := g_7;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        return
};

spec "WP::<T>::first_mut"(pLft_a, pty_T0, self)
[[ "<&mut WP<T> as gilogic::Ownable>::own"(#pLft_a, #pty_T0, #self)
  * <lft>(#pLft_a; true)
  * types(#lvar_0: Obj, #lvar_1: List) * (#self == {{ #lvar_0, #lvar_1 }})
  * (self == #self)
  * (pty_T0 == #pty_T0)
  * (pLft_a == #pLft_a) ]]
[[ "<&mut T as gilogic::Ownable>::own"(#pLft_a, #pty_T0, ret)
  * <lft>(#pLft_a; true) ]]
normal

proc "WP::<T>::first_mut"(pLft_a, pty_T0, self) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, g_0);
        m_2 := [alloc]({{ "ref", true, pty_T0 }});
        m_3 := [alloc]({{ "ref", true, pty_T0 }});
        m_6ret := [alloc]({{ "ref", true, pty_T0 }});
  bb0:  g_2 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_1 := l-nth(g_2, 0i);
        g_3 := [new_lft]();
        pLft_13 := l-nth(g_3, 0i);
        m_5 := g_1;
        g_4 := m_5;
        apply "freeze_xy::<T>"(pLft_a,pty_T0,g_4);
        goto bb1;
  bb1:  g_6 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_5 := l-nth(g_6, 0i);
        g_8 := [load_value](l-nth(g_5, 0i), l+(l-nth(g_5, 1i), {{ {{ "f", 0i, {{ "adt", "WP", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_7 := l-nth(g_8, 0i);
        g_9 := [new_lft]();
        pLft_14 := l-nth(g_9, 0i);
        u := [store_value](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }}, {{ l-nth(g_7, 0i), l+(l-nth(g_7, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ pty_T0 }} }} }} }}) }});
        g_11 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_10 := l-nth(g_11, 0i);
        g_12 := [new_lft]();
        pLft_15 := l-nth(g_12, 0i);
        m_8 := g_10;
        g_13 := m_8;
        g_14 := "extract_x::<T>"(pLft_a, pty_T0, g_13);
        m_7 := g_14;
        goto bb2;
  bb2:  g_16 := [load_value](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }}, true);
        g_15 := l-nth(g_16, 0i);
        g_17 := [new_lft]();
        pLft_12 := l-nth(g_17, 0i);
        u := [store_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }}, g_15);
        g_19 := [load_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }}, true);
        g_18 := l-nth(g_19, 0i);
        g_20 := [new_lft]();
        pLft_11 := l-nth(g_20, 0i);
        u := [store_value](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }}, g_18);
        g_22 := [load_value](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }}, true);
        g_21 := l-nth(g_22, 0i);
        g_23 := [new_lft]();
        pLft_9 := l-nth(g_23, 0i);
        m_0 := g_21;
        g_24 := m_0;
        ret := g_24;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }});
        u := [free](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }});
        u := [free](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }});
        return
};

spec "WP::<T>::second_mut"(pLft_a, pty_T0, self)
[[ "<&mut WP<T> as gilogic::Ownable>::own"(#pLft_a, #pty_T0, #self)
  * <lft>(#pLft_a; true)
  * types(#lvar_0: Obj, #lvar_1: List) * (#self == {{ #lvar_0, #lvar_1 }})
  * (self == #self)
  * (pty_T0 == #pty_T0)
  * (pLft_a == #pLft_a) ]]
[[ "<&mut T as gilogic::Ownable>::own"(#pLft_a, #pty_T0, ret)
  * <lft>(#pLft_a; true) ]]
normal

proc "WP::<T>::second_mut"(pLft_a, pty_T0, self) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, g_0);
        m_2 := [alloc]({{ "ref", true, pty_T0 }});
        m_3 := [alloc]({{ "ref", true, pty_T0 }});
        m_6ret := [alloc]({{ "ref", true, pty_T0 }});
  bb0:  g_2 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_1 := l-nth(g_2, 0i);
        g_3 := [new_lft]();
        pLft_13 := l-nth(g_3, 0i);
        m_5 := g_1;
        g_4 := m_5;
        apply "freeze_xy::<T>"(pLft_a,pty_T0,g_4);
        goto bb1;
  bb1:  g_6 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_5 := l-nth(g_6, 0i);
        g_8 := [load_value](l-nth(g_5, 0i), l+(l-nth(g_5, 1i), {{ {{ "f", 1i, {{ "adt", "WP", {{ pty_T0 }} }} }} }}), {{ "ptr", true, {{ "adt", "N", {{ pty_T0 }} }} }}, true);
        g_7 := l-nth(g_8, 0i);
        g_9 := [new_lft]();
        pLft_14 := l-nth(g_9, 0i);
        u := [store_value](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }}, {{ l-nth(g_7, 0i), l+(l-nth(g_7, 1i), {{ {{ "f", 0i, {{ "adt", "N", {{ pty_T0 }} }} }} }}) }});
        g_11 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }}, true);
        g_10 := l-nth(g_11, 0i);
        g_12 := [new_lft]();
        pLft_15 := l-nth(g_12, 0i);
        m_8 := g_10;
        g_13 := m_8;
        g_14 := "extract_y::<T>"(pLft_a, pty_T0, g_13);
        m_7 := g_14;
        goto bb2;
  bb2:  g_16 := [load_value](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }}, true);
        g_15 := l-nth(g_16, 0i);
        g_17 := [new_lft]();
        pLft_12 := l-nth(g_17, 0i);
        u := [store_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }}, g_15);
        g_19 := [load_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }}, true);
        g_18 := l-nth(g_19, 0i);
        g_20 := [new_lft]();
        pLft_11 := l-nth(g_20, 0i);
        u := [store_value](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }}, g_18);
        g_22 := [load_value](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }}, true);
        g_21 := l-nth(g_22, 0i);
        g_23 := [new_lft]();
        pLft_9 := l-nth(g_23, 0i);
        m_0 := g_21;
        g_24 := m_0;
        ret := g_24;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "WP", {{ pty_T0 }} }} }});
        u := [free](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "ref", true, pty_T0 }});
        u := [free](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, pty_T0 }});
        u := [free](l-nth(m_6ret, 0i), l-nth(m_6ret, 1i), {{ "ref", true, pty_T0 }});
        return
};

